### PR TITLE
Load blink IDL

### PR DESCRIPTION
**Pull request by mdittmer**
@arobins and @kgrgreer PTAL.

- Models for IDL files from different APIs (Gitiles and Github)
- DAO for Blink IDL files; has async init using Chromium code search to get file listing
- Store-and-forward DAO to provide standard DAO interface for async init DAO(s)
- Tests for DAO implementations

### CODE REVIEW

**Comment by mdittmer** [comment]
Just realized that I meant to include more tests in this PR. There aren't enough negative tests for store-and-forward (i.e., should reject on errors that are not retryable). Also meant to write store-and-forward + IDL file DAO integration tests. We can take care of these in a follow-up PR.

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  requires: [
+    'foam.dao.InternalException',
+    'foam.dao.MDAO',
+    'foam.net.HTTPRequest',
+    'org.chromium.webidl.GitilesIDLFile',
+  ],
+
+  constants: {
+    branchToCommitURL: 'https://chromium.googlesource.com/chromium/src/+/master?format=JSON',
+    codeSearchQueryURL: 'https://cs.chromium.org/codesearch/json/search_request:1?search_request=b&query=file%3Athird_party%2FWebKit%2FSource.*idl%24+-file%3A%2Ftesting%2F+-file%3A%2Fbindings%2Ftests%2F&max_num_results=2000&internal_options=b&internal_options=e&search_request=e',
```
Comment (arobins) [inline lib/org/chromium/webidl/BlinkIDLFileDAO.js]: 2000 IDL files should be enough for anyone?

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    },
+    {
+      class: 'String',
+      documentation: 'Git commit hash that DAO points to.',
+      name: 'commit',
+    },
+    {
+      class: 'Boolean',
+      documentation: `Internal state: Is the DAO ready to accept operations?
+        May throw a faom.dao.InternalException. To automatically catch retry
```
Comment (arobins) [inline lib/org/chromium/webidl/BlinkIDLFileDAO.js]: nit: typo in foam

File: lib/org/chromium/webidl/BlinkIDLFile.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+foam.CLASS({
+  package: 'org.chromium.webidl',
+  name: 'BlinkIDLFile',
+  extends: 'org.chromium.webidl.GitilesIDLFile',
+
+  documentation: 'An IDL file accessible via Gitiles.',
```
Comment (arobins) [inline lib/org/chromium/webidl/BlinkIDLFile.js]: A Blink IDL file?

File: lib/org/chromium/webidl/GithubIDLFile.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+foam.CLASS({
+  package: 'org.chromium.webidl',
+  name: 'GithubIDLFile',
+  extends: 'org.chromium.webidl.IDLFile',
+
+  documentation: 'An IDL file accessible via Gitiles.',
```
Comment (arobins) [inline lib/org/chromium/webidl/GithubIDLFile.js]: github

File: lib/org/chromium/webidl/IDLFile.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+foam.CLASS({
+  package: 'org.chromium.webidl',
+  name: 'IDLFile',
+
+  documentation: 'An IDL file.',
```
Comment (arobins) [inline lib/org/chromium/webidl/IDLFile.js]: Mention fetchable over HTTP.

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js (hunk 1 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/BlinkIDLFileDAO.js]: That's right. We're south of 1200 right now.

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js (hunk 2 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/BlinkIDLFileDAO.js]: Done.

File: lib/org/chromium/webidl/BlinkIDLFile.js (hunk 3 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/BlinkIDLFile.js]: True. Turns out this file was a mistake. I was going to add some Blink-specific stuff but decided against it. Mistakenly got roped in with other files in the dir though.

File: lib/org/chromium/webidl/GithubIDLFile.js (hunk 4 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/GithubIDLFile.js]: Done.

File: lib/org/chromium/webidl/IDLFile.js (hunk 5 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/IDLFile.js]: Done.

**Comment by mdittmer** [comment]
@arobins PTAL at updates. I think this is ready to go.

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js (hunk 1 above)
Comment (arobins) [reply lib/org/chromium/webidl/BlinkIDLFileDAO.js]: I think you should extract the max count into a constant.

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+  methods: [
+    function init() {
+      this.branchToCommit_().then(this.onCommitSet);
+    },
+    function put(obj) {
+      var notReady = this.checkNotReady_();
+      if (notReady) return notReady;
+
+      return this.delegate.put(obj);
```
Comment (arobins) [inline lib/org/chromium/webidl/BlinkIDLFileDAO.js]: In all of these, you could write:

return this.checkNotReady_() || this.delegate.put(obj);

**Review by arobins — APPROVED** [review]

File: lib/org/chromium/webidl/DAOOperation.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      documentation: 'Arguments object associated with operation.',
+      name: 'args',
+    },
+    {
+      name: 'promise_',
+      factory: function() {
+        var self = this;
+        var resolve;
+        var reject;
+        var promise = new Promise(function(res, rej) {
```
Comment (kgrgreer) [inline lib/org/chromium/webidl/DAOOperation.js:26]: We should make this pattern into a reusable 'Future' object. (Not now, just more of a note to myself.)

**Comment by kgrgreer** [comment]
LGTM

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js (hunk 1 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/BlinkIDLFileDAO.js]: Done.

File: lib/org/chromium/webidl/BlinkIDLFileDAO.js (hunk 6 above)
Comment (mdittmer) [reply lib/org/chromium/webidl/BlinkIDLFileDAO.js]: Done (here and elsewhere).

**Review by mdittmer — COMMENTED** [review]
Addressed comments from @arobins

**Comment by mdittmer** [comment]
Towards #4

**Pull request merged**